rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.vectors/90_knn_query_with_filter.yml

```diff
@@ -0,0 +1,253 @@
+  - match: { hits.hits.1.fields.my_name.0: v2 }
+  - match: { hits.hits.2._id: "6" }
+  - match: { hits.hits.2.fields.my_name.0: v2 }
+
+---
+"PRE_FILTER: pre-filter across multiple aliases":
+  - do:
+      search:
+        index: my_alias,my_alias1
+        body:
+          size: 6
+          fields: [ my_name ]
+          query:
+            knn:
+              field: my_vector
+              query_vector: [1, 1, 1, 1]
+              num_candidates: 100
+
+  - match: { hits.total.value: 10 } # 5 docs from each alias
+  - length: {hits.hits: 6}
+  - match: { hits.hits.0._id: "1" }
+  - match: { hits.hits.0.fields.my_name.0: v1 }
+  - match: { hits.hits.1._id: "2" }
+  - match: { hits.hits.1.fields.my_name.0: v2 }
+  - match: { hits.hits.2._id: "3" }
+  - match: { hits.hits.2.fields.my_name.0: v1 }
+  - match: { hits.hits.3._id: "4" }
+  - match: { hits.hits.3.fields.my_name.0: v2 }
+  - match: { hits.hits.4._id: "5" }
+  - match: { hits.hits.4.fields.my_name.0: v1 }
+  - match: { hits.hits.5._id: "6" }
+  - match: { hits.hits.5.fields.my_name.0: v2 }
+
+---
+"PRE_FILTER: knn query with internal filter as pre-filter":
+  - do:
+      search:
+        index: my_index
+        body:
+          size: 3
+          fields: [ my_name ]
+          query:
+            knn:
+              field: my_vector
+              query_vector: [1, 1, 1, 1]
+              num_candidates: 5
+              filter:
+                term:
+                  my_name: v2
+
+  - match: { hits.total.value: 5 }
+  - length: {hits.hits: 3}
+  - match: { hits.hits.0._id: "2" }
+  - match: { hits.hits.0.fields.my_name.0: v2 }
+  - match: { hits.hits.1._id: "4" }
+  - match: { hits.hits.1.fields.my_name.0: v2 }
+  - match: { hits.hits.2._id: "6" }
+  - match: { hits.hits.2.fields.my_name.0: v2 }
+
+---
+"POST_FILTER: knn query with filter from a parent bool query as post-filter":
+  - do:
+      search:
+        index: my_index
+        body:
+          size: 3
+          fields: [ my_name ]
+          query:
+            bool:
+              must:
+                - term:
+                    my_name: v2
+                - knn:
+                    field: my_vector
+                    query_vector: [1, 1, 1, 1]
+                    num_candidates: 5
+
+  - match: { hits.total.value: 2 }
+  - length: {hits.hits: 2} # knn query returns top 5 docs, but they are post-filtered to 2 docs
+  - match: { hits.hits.0._id: "2" }
+  - match: { hits.hits.0.fields.my_name.0: v2 }
+  - match: { hits.hits.1._id: "4" }
+  - match: { hits.hits.1.fields.my_name.0: v2 }
+
+  - do:
+      search:
+        index: my_index
+        body:
+          size: 3
+          fields: [ my_name ]
+          query:
+            bool:
+              must:
+                - term:
+                    my_name: v2
+                - knn:
+                    field: my_vector
+                    query_vector: [ 1, 1, 1, 1 ]
+                    num_candidates: 5
+                    filter:
```

Comment:
Speaking purely from the viewpoint of a consumer of this API without deep knowledge of the underlying implementation of kNN search in ES:

In the spirit of simplifying the API, is there an opportunity to a) remove the 'filter' field from the knn query and b) have the kNN query be the post-filter instead?  

Assuming the current spec changes, if I were to provide a 'term' and a 'knn' clause then I am actually more likely to see fewer results than if I were to provide a 'term' clause and a 'knn' clause with a 'filter', due to kNN trying to satisfy the 'size' requirement as quickly as possible.

More clauses yielding more results feels counter-intuituive in a context other than 'should'

Reply:
@mshameti Thanks for your feedback.

Are you suggesting to  always exercise filters for knn query as post filters?
Pre-filtering is very important aspect of knn query that ensures you will get some results, while post-filter can eliminate all results.

> 'term' and a 'knn' clause then I am actually more likely to see fewer results than if I were to provide a 'term' clause and a 'knn' clause with a 'filter', due to kNN trying to satisfy the 'size' requirement as quickly as possible.

Not sure how is it more likely?  It is actually less likely. `knn` with a `filter` returns the same or less number of results than knn query, but never more.

Reply:
Thanks for looking! 

I've created a Gist to demonstrate this a bit better hopefully. I've listed out my assumptions in there as well.

 https://gist.github.com/mshameti/774639631363457dec310deee4f15766

Reply:
> More clauses yielding more results feels counter-intuitive in a context other than 'should'

It doesn't return more results. `knn` will apply the filter and when combined with other queries, those may be "post filtered" out. The doc count hit may still be `k` as we found there are at least `k` documents that match that filter and scored them.

But the scored hits returned must satisfy all query provided filters. I am not sure how it returns MORE results. Reading through the gist, I am still not 100% sure I understand the concern.

Could you clarify what you mean by "More clauses yielding more results feels counter-intuitive in a context other than 'should'"? 

The scenario with a `must`

```
GET /knn-test/_search
{
  "query": {
    "bool": {
      "must": [
        {
          "match": {
            "author": "Martin"
          }
        },
        {
          "knn": {
            "query_vector": [1, 1, 1, 1],
            "field": "paragraph_embedding",
            "num_candidates": 1,
            "filter": {
              "match": {
                "author": "Martin"
              }
            }
          }
        }
      ]
    }
  },
  "size": 1
}
```

This would result in only documents that satisfy 
```
{
          "match": {
            "author": "Martin"
          }
        }
```

And `knn` will only score documents where
```
"filter": {
              "match": {
                "author": "Martin"
              }
            }
```

Is true.

So, it isn't finding MORE documents. It is just scoring different documents, which still matches the typical way a query works, its just that other queries have a "natural pre-filter" which is the matching terms :).

Reply:
> a. treat the top level match as a pre-filter

I understand now. 

We considered that, but it gets very difficult to get correct and would have non-obvious edge cases. 

What do you do in a multiple nested `bool` query?

What if a higher level `bool` has a `knn` query? 

Do we ignore `should` clauses?

We may in the future consider making the filter optional and dynamically applying the correct or expected pre-filter.

Reply:
Great questions! For future use, I created a Gist of the situtations you mentioned and a couple others to help me reason through this. https://gist.github.com/mshameti/2b87802869d83df6b06abade142a922e

I most likely have not covered all the bases but if you do decide to revisit this it's here for future use. 

Thanks again for engaging in the discussion, and as the proposed API change is already an improvement to the existing API, please consider this conversation resolved from my end!